mautic/mautic
# [DPMMA-1719] Don't track bot email clicks or opens

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [❌]
| New feature/enhancement? (use the a.x branch)      | [✅]
| Deprecations?                          | [❌]
| BC breaks? (use the c.x branch)        | [❌]
| Automated tests included?              | [✅] 
| Related user documentation PR URL      |
| Related developer documentation PR URL | 
| Issue(s) addressed                     | 

#### Description:

This PR try to extends bot hit filtering from [PR#3501](https://github.com/mautic/mautic/pull/3501) to exclude also email clicks and opens. Before registering the statistics, the IP address is verified. If it is not trackable, e.g. due to a user agent that contains bot keywords, statistics are not saved.

#### Steps to test this PR:

1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Go to **Configuration** -> **System Settings** -> **Miscellaneous Settings**. 
3. In **List of Bots to not track with (one per line)**, enter the keywords that will be searched in the request user agent. 
4. Switch **Exclude Bots from email statistics** to **Yes**. 
![email_bots](https://user-images.githubusercontent.com/102536220/229764963-ff9692ec-486c-463d-87ab-a3cae9facfbd.jpg)
6. Create a sample email and send it to your mailbox.
7. Open email, download the pictures and click on one of the links in the mail client whose User Agent fragment you entered earlier on the **List of Bots to not track** list, e.g. Thunderbird. With the **Exclude Bots from email statistics** option enabled, the opening or clicks should not be recorded.
